unit/optional.cpp

```diff
@@ -0,0 +1,37 @@
+/*******************************************************************\
+
+ Module: nonstd::optional unit tests
+
+ Author: Diffblue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <catch.hpp>
+#include <util/optional.h>
+
+TEST_CASE("Optional without a value")
+{
+  optionalt<bool> maybe_value;
+  REQUIRE(maybe_value.has_value()==false);
+  try
+  {
+    maybe_value.value();
+    REQUIRE(false); // Should have thrown an exception
+  }
+  catch(const bad_optional_accesst&) { }
```

Comment:
Catch provides the ability to verify exceptions: https://github.com/philsquared/Catch/blob/master/docs/assertions.md#exceptions